### ISSUE TITLE
A way for schema plugins to extend Query class

### ISSUE THREAD

**Issue by tsabolov** [Automattic/mongoose#2696]
The problem is that sometimes it's needed to have own methods to the conditions chain.
I.e. I have a plugin that adds access permissions to the schema (call it `acl`). It adds an instance method `findByOwner` to the schema.
Also I have a plugin that adds flags to the same schema (call it `flags`). It adds an instance method `findByFlag` to the schema.
Now I want to query documents of this schema filtered by the owner as well as by some flag. Each of these filters could be applied separately:
```js
MyModel.findByOwner(owner, callback);
```
and
```js
MyModel.findByFlag(flag, callback);
```
But it's impossible to chain them like:
```js
MyModel.findByOwner(owner).findByFlag(flag).exec(callback);
```
Someone advised to extend `mongoose.Query`'s prototype. But IMO it breaks the principle that plugin only affects the schema it was applied to.

Maybe it's possible to move Query class reference to the Schema and call it in the model methods like:
```js
Model.where = function where (path, val) {
  var mq = new this.schema.Query({}, {}, this, this.collection).find({});
  return mq.where.apply(mq, arguments);
};
```
Plugins will be able to overwrite Query class reference with the child of `mongoose.Query` extended with the necessary methods.

Are there any pitfalls with this solution?

**Comment by vkarpov15** [comment]
Worth investigating, but I'm not sure there isn't already a good solution though. What's wrong with extending `Query`? It would not necessarily affect any other schemas - just have a new `MyModelQuery` class whose prototype is an `Object.create(Query)` and then add functions to your heart's content. Query functions return `this`, so I can't quite think of any problems with that approach.

**Comment by tsabolov** [comment]
Nice solution, btw. Didn't think about it that way. But anyway I need to overwrite `mongoose.Model`'s methods to return instances of my own `mongoose.Query` descent.
Thank you.

**Comment by vkarpov15** [comment]
Yeah you're right, that's a little messy. I'll mark this ticket for a future release - not sure where on the roadmap this'll fall, but I see your point.

**Comment by xpepermint** [comment]
+1

**Comment by vkarpov15** [comment]
Released in 4.5, see #3740, etc.
